lib/galaxy/jobs/__init__.py

```diff
@@ -1721,14 +1706,13 @@ def fail():
         self.tool.call_hook('exec_after_process', self.app, inp_data=inp_data,
                             out_data=out_data, param_dict=param_dict,
                             tool=self.tool, stdout=job.stdout, stderr=job.stderr)
-        job.command_line = unicodify(self.command_line)
 
         collected_bytes = 0
         # Once datasets are collected, set the total dataset size (includes extra files)
         for dataset_assoc in job.output_datasets:
             if not dataset_assoc.dataset.dataset.purged:
-                dataset_assoc.dataset.dataset.set_total_size()
-                collected_bytes += dataset_assoc.dataset.dataset.get_total_size()
+                # don't call get_total_size - forces a flush we don't want in here.
+                collected_bytes += dataset_assoc.dataset.dataset.set_total_size()
```

Comment:
```suggestion
                collected_bytes += dataset_assoc.dataset.set_total_size()
```

Reply:
Both versions are correct. job.output_datasets are `JobToOutputDatasetAssociation` objects. So `dataset_assoc.dataset` is the HDA and `dataset_assoc.dataset.dataset` is a `Dataset`. I guess it saves one function call to use `dataset_assoc.dataset.dataset.set_total_size()` ?

Reply:
My reasoning is that the HDA is correct level of abstraction here, unless there is an important performance gain in skipping a method call.

Reply:
That is fair - I mean this is performant critical code to my mind but the extra python function call is the least of our problems so I'll use the better abstraction.

Reply:
But the size refers to the dataset, not the history dataset relation ? In my mind we'll have to separate out extra files at one point in the future and relate that back to the dataset, so that we can have HDAs of different datatypes sharing a dataset.

Reply:
I don't think I agree with that - at least at first glance. The dataset is physical collection of files. I think the right way to handle different composite files between HDAs would be to duplicate the dataset with different contents and find some other mechanism to de-duplicate actual files? A concept below Dataset maybe?

Reply:
I guess put another way Metadata files and extra files made very different choices and should be converged in some fashion - but toward datasets would make more sense to me and my view of things. I could be wrong though.

Reply:
But even if we had a concept below Dataset the file size wouldn't belong on the HDA but the Dataset? It seems counterintuitive to me for the relation to have a size, especially since the composite file content is also part of the Dataset. Effectively we get and set the size of the `Dataset` here, so I'm fine with merging it as is, but I feel like the current form is clearer both to what actually happens and how we may want to go forward ?

Reply:
I didn't see your second comment before posting my last one. I guess we can revisit this when someone starts working on it, let's get this in. MetadataFile being part of the HDA seems to make more sense to me, I didn't think about that. In fact making extra files a first class object that relates to HDAs would probably also neatly close that gap I am worried about in our model.